dask/array/core.py

```diff
@@ -1583,23 +1639,17 @@ def elemwise(op, *args, **kwargs):
     if not set(['name', 'dtype']).issuperset(kwargs):
         raise TypeError("%s does not take the following keyword arguments %s" %
             (op.__name__, str(sorted(set(kwargs) - set(['name', 'dtype'])))))
-    out_ndim = max(len(arg.shape) if isinstance(arg, Array) else 0
-                   for arg in args)
+
+    out_ndim = max(len(getattr(arg, 'shape', ())) for arg in args)
     expr_inds = tuple(range(out_ndim))[::-1]
 
-    arrays = [arg for arg in args if isinstance(arg, Array)]
-    other = [(i, a) for i, a in enumerate(args) if not isinstance(a, Array)]
-    if any(isinstance(arg, np.ndarray) for arg in args):
-        raise NotImplementedError("Dask.array operations only work on dask "
-                                  "arrays, not numpy arrays.")
     if 'dtype' in kwargs:
         dt = kwargs['dtype']
-    elif not all(a._dtype is not None for a in arrays):
+    elif any(a._dtype is None for a in args if isinstance(a, Array)):
         dt = None
     else:
-
         vals = [np.empty((1,) * a.ndim, dtype=a.dtype)
```

Comment:
Note that this is not really the right logic here. The dtype signature is fixed independent of the number of dimensions except for 0d arrays, for which the dtype depends on the value according to some complex set of rules: https://github.com/numpy/numpy/issues/6240